### PR TITLE
BB2-434: Amazon Linux 2 + Python 3.7 migration

### PR DESCRIPTION
**Pull request by njdister**
**JIRA Ticket:**
[BB2-434](https://jira.cms.gov/browse/BB2-434)

**User Story or Bug Summary:**
Migrate AMIs to use Amazon Linux 2 and Python 3.7

### What Does This PR Do?
This PR takes a bit of a heavy handed approach to duplicating existing deployment code in order to offer an extremely simple rollback or side-by-side deployment strategy via `platinum_source` feature flag:
- Set up new Jenkins jobs for Amazon Linux 2 that use this branch where `platinum_source == "amzn2"`
- Once ready to migrate, merge, set the Amazon Linux 2 jobs to master branch, and disable the old Jenkins jobs.
- If a rollback is required, create a branch where `platinum_source == "rhel7"` (or whatever since it doesn't matter), and enable the old Jenkins jobs with that branch in order to build AMIs from RHEL 7 again.

I have duplicated the Jenkinsfiles, packer definitions, and top level playbooks for AMI builds in order to support the outlined rollback strategy, and will add comments where appropriate regarding changes.  I have also created a new role, `amzn2_base`, as a LOT of work in the old platinum build was unnecessary for Amazon Linux 2.  Only a few minor things in the app build roles have changed, particularly around uwsgi since its plugin for Python 3.7 must be installed from pip rather than system package.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check these things, in particular:

- Is this a good rollback strategy? Can you think of something better?
- How important is updating the local development stack to 3.7?  So far I have tested in dev, test, and sandbox with real traffic via a connected load balancer and not noticed any errors, but I imagine this would be a nice to do before this is merged for safety of future local development.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then here's a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence: N/A.
    * Does this PR add any new software dependencies? **Yes**.
    * Does this PR modify or invalidate any of our security controls? **No**.
    * Does this PR store or transmit data that was not stored or transmitted before? **No**.
* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR should not be merged unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons? **Yes**.

Requires an SIA approval before merge!

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following pre-requisite changes have been fully deployed:

- Python 3.7 wheel for PyYAML: https://github.com/CMSgov/bluebutton-web-server/pull/884

### Submitter Checklist

I have gone through and verified that...:

* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] I have named this PR and its branch such that they'll be automatically be linked to the (most) relevant Jira issue, per: <https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html>.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. BFD, SLS, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: Jenkinsfiles/Jenkinsfile.cbc-build-app-ami-amzn2 [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    AWS_SECRET_ACCESS_KEY = credentials("bb2-aws-key-secret")
+    AWS_DEFAULT_REGION = "us-east-1"
+    VAULT_PASSWORD = credentials("bb2-vault-password")
+  }
+
+  parameters {
+    string(
+      name: "APP_REPO_BRANCH",
+      // FIXME: Temporary value for testing, revert to master prior to merging!
+      defaultValue: "njdister/BB2-434-amzn2-migration",
```
Comment (njdister) [inline Jenkinsfiles/Jenkinsfile.cbc-build-app-ami-amzn2]: Temporary value, revert prior to merge!

File: Jenkinsfiles/Jenkinsfile.cbc-build-app-ami-amzn2 [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      description: "The branch of the application repo to build in AMI. Defaults to master."
+    )
+
+    string(
+      name: "DEPLOY_REPO_BRANCH",
+      // FIXME: Temporary value for testing, revert to master prior to merging!
+      defaultValue: "njdister/BB2-434-amzn2-migration",
```
Comment (njdister) [inline Jenkinsfiles/Jenkinsfile.cbc-build-app-ami-amzn2]: Temporary value, revert prior to merge!

File: Jenkinsfiles/Jenkinsfile.cbc-build-app-ami-amzn2 [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    stage("Determine Source AMI") {
+      steps {
+        script {
+          if (params.PLATINUM_AMI_ID != "") {
+            echo("Setting SOURCE_AMI from PLATINUM_AMI_ID")
+            SOURCE_AMI = params.PLATINUM_AMI_ID
+          } else {
+            echo("Setting SOURCE_AMI from latest platinum built in AWS")
+            SOURCE_AMI = sh(returnStdout: true, script: "aws ec2 describe-images \
+              --filters 'Name=name,Values=bb2-amzn2-platinum-*' \
```
Comment (njdister) [inline Jenkinsfiles/Jenkinsfile.cbc-build-app-ami-amzn2:56]: Default search filter changed for new Amazon Linux 2 AMI naming convention.

File: Jenkinsfiles/Jenkinsfile.cbc-build-app-ami-amzn2 [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    stage("Build App AMI") {
+      steps {
+        sh """
+          packer build -color=false \
+            -var 'vault_password_file=${VAULT_PASSWORD}' \
+            -var 'source_ami=${SOURCE_AMI}' \
+            -var 'release_version=${APP_REPO_BRANCH}' \
+            -var 'git_branch=${APP_REPO_BRANCH}' \
+            -var 'deployment_branch=${DEPLOY_REPO_BRANCH}' \
+            packer/build_app_ami_amzn2.json
```
Comment (njdister) [inline Jenkinsfiles/Jenkinsfile.cbc-build-app-ami-amzn2:73]: Packer definition changed to new Amazon Linux 2 definition.

File: Jenkinsfiles/Jenkinsfile.cbc-build-platinum-ami-amzn2 [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    AWS_ACCESS_KEY_ID = credentials("bb2-aws-key-id")
+    AWS_SECRET_ACCESS_KEY = credentials("bb2-aws-key-secret")
+    AWS_DEFAULT_REGION = "us-east-1"
+    VAULT_PASSWORD = credentials("bb2-vault-password")
+  }
+
+  parameters {
+    string(
+      name: "GOLD_AMI_SEARCH_FILTER",
+      defaultValue: "amzn2legacy*",
```
Comment (njdister) [inline Jenkinsfiles/Jenkinsfile.cbc-build-platinum-ami-amzn2:19]: Default search filter changed to find latest Amazon Linux 2 legacy (aka. not greenfield) gold image.

File: Jenkinsfiles/Jenkinsfile.cbc-build-platinum-ami-amzn2 [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+      }
+    }
+
+    stage("Build Platinum AMI") {
+      steps {
+        sh """
+          packer build -color=false \
+            -var 'vault_password_file=${VAULT_PASSWORD}' \
+            -var 'source_ami=${SOURCE_AMI}' \
+            packer/build_platinum_ami_amzn2.json
```
Comment (njdister) [inline Jenkinsfiles/Jenkinsfile.cbc-build-platinum-ami-amzn2:65]: Packer definition changed to new Amazon Linux 2 definition.

File: packer/build_app_ami_amzn2.json [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  },
+  "builders": [{
+    "type": "amazon-ebs",
+    "ssh_username": "ec2-user",
+    "instance_type": "m4.large",
+    "region": "us-east-1",
+    "subnet_id": "{{user `subnet_id`}}",
+    "associate_public_ip_address": "false",
+    "source_ami": "{{user `source_ami`}}",
+    "ami_name": "bb2-amzn2-{{user `git_branch`}}-{{isotime \"20060102030405\"}}",
```
Comment (njdister) [inline packer/build_app_ami_amzn2.json:18]: Updated AMI naming convention

File: packer/build_app_ami_amzn2.json [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    "ssh_pty": true,
+    "tags": {
+        "Name": "bb2-amzn2-{{user `git_branch`}}-{{isotime \"20060102030405\"}}",
```
Comment (njdister) [inline packer/build_app_ami_amzn2.json:21]: Updated AMI naming convention

File: packer/build_app_ami_amzn2.json [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        "Application": "BlueButton",
+        "Function": "AppServer",
+        "Layer":  "APP",
+        "Release": "{{user `release_version`}}"
+    }
+  }],
+    "provisioners": [
+        {
+            "type": "ansible",
+            "playbook_file": "playbook/build_app_ami_amzn2/main.yml",
```
Comment (njdister) [inline packer/build_app_ami_amzn2.json:31]: Using new top level playbook definition for Amazon Linux 2

File: packer/build_platinum_ami_amzn2.json [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  },
+  "builders": [{
+    "type": "amazon-ebs",
+    "ssh_username": "ec2-user",
+    "instance_type": "m4.large",
+    "region": "us-east-1",
+    "subnet_id": "{{user `subnet_id`}}",
+    "associate_public_ip_address": "false",
+    "source_ami": "{{user `source_ami`}}",
+    "ami_name": "bb2-amzn2-platinum-{{isotime \"20060102030405\"}}",
```
Comment (njdister) [inline packer/build_platinum_ami_amzn2.json:15]: Updated AMI naming convention

File: packer/build_platinum_ami_amzn2.json [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    "launch_block_device_mappings": [
+      {
+        "device_name": "/dev/xvda",
+        "volume_type": "gp2",
+        "volume_size": 30,
+        "delete_on_termination": true
+      }
+    ],
+    "tags": {
+      "Name": "bb2-amzn2-platinum-{{isotime \"20060102030405\"}}"
```
Comment (njdister) [inline packer/build_platinum_ami_amzn2.json:26]: Updated AMI naming convention

File: packer/build_platinum_ami_amzn2.json [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    }
+  }],
+    "provisioners": [
+        {
+            "type": "ansible",
+            "playbook_file": "playbook/build_platinum_ami_amzn2/main.yml",
```
Comment (njdister) [inline packer/build_platinum_ami_amzn2.json:32]: Using new top level playbook definition for Amazon Linux 2

File: packer/build_platinum_ami_amzn2.json [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+    "source_ami": "{{user `source_ami`}}",
+    "ami_name": "bb2-amzn2-platinum-{{isotime \"20060102030405\"}}",
+    "ssh_pty": true,
+    "launch_block_device_mappings": [
+      {
+        "device_name": "/dev/xvda",
+        "volume_type": "gp2",
+        "volume_size": 30,
+        "delete_on_termination": true
+      }
```
Comment (njdister) [inline packer/build_platinum_ami_amzn2.json:23]: Amazon Linux 2 gold image has a 8GB root volume by default, we need to expand that for log space.

**Comment by njdister** [comment]
This is tentatively approved pending Nessus scans of EC2 instances in our lower environments (where I have deployed AMZN2 AMIs) so I'm opening this up for review.

Edit: This has passed scans and SIA, so this is clear for deployment when reviewed.

**Review by JFU-GIT — APPROVED** [review]
Thanks for the demo, it's a good move considering the many benefits.

**Review by dtisza1 — APPROVED** [review]
This looks good to me!

Thank you for the self-review comments and the zoom call earlier. This made the review easier!

**Pull request merged**